### DISCUSSION TITLE
I can't get my head around this

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [ispc/ispc#2859] [General]
I've read this part of the documentation multiple times and some of it still doesn't make sense to me.

> The variability of template type parameter T may be overwritten by uniform and varying keywords, so uniform T and varying T are always valid types. But when uniform T and varying T are used to specify template function parameters it has an effect on template argument deduction process. If variability keyword is specified and the type was successfully deduced, the default variability of type T is assumed to be the opposite of the variability keyword. The logic behind it is that the keyword was specified "on purpose" to changed the variability of the type T. Consider the following example:
```cpp
template <typename T> void foo1(T t);
template <typename T> void foo2(uniform T t);
template <typename T> void foo3(varying T t);

void bar() {
    uniform int ui;
    varying int vi;
    foo1(ui); // T is uniform int (makes perfect sense)
    foo1(vi); // T is varying int (makes perfect sense)
    foo2(ui); // T is varying int! (why varying?)
    foo2(vi); // error: varying type cannot be passed to uniform parameter   (makes sense)
    foo3(ui); // T is uniform int! (ok so its like we are overriding the variability here, makes sense)
    foo3(vi); // T is uniform int! (what??)
}
```

I have commented on what makes sense to me and what doesn't. I appreciate if some one could clarify this for me

**Comment by MkazemAkhgary** [comment]
I think I now understand what is happening. in case of `foo3(vi)`, the type `T` is uniform but the parameter `t` is still varying. now it makes sense.

```cpp
foo2(ui); // parameter t is uniform but type T is varying.
foo2(vi); // parameter t is uniform hence the error (although type T is varying)
foo3(ui); // parameter t is varying but type T is uniform
foo3(vi); // parameter t is varying but type T is uniform
```

is my understanding correct now?

**Reply by aneshlya** [reply]
Yes! You can find more explanations in the design document: https://github.com/ispc/ispc/blob/main/docs/design/templates.rst#L31
